File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
    void genrate(int n,int left,int right,vector<string>&ans,string &temp){
        if(left+right==2*n){
            ans.push_back(temp);
            return;
        }

        //left k liye
        if(left<n){
            temp.push_back('(');
            genrate(n,left+1,right,ans,temp);
            temp.pop_back();
        }
        if(right<left){
            temp.push_back(')');
            genrate(n,left,right+1,ans,temp);
            temp.pop_back();
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        
        vector<string>ans;
        string temp;
        genrate(n,0,0,ans,temp);
        return ans;
        
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        st=[]
        res=[]
        def sol(op,cl):
            if op==cl==n:
                res.append("".join(st))
                return
            if op<n:
                st.append('(')
                sol(op+1,cl)
                st.pop()
            if cl<op:
                st.append(')')
                sol(op,cl+1)
                st.pop()

        sol(0,0)
        return res